src/shared/clients/defaultS3Client.ts

```diff
@@ -0,0 +1,328 @@
+                    name: s3Bucket.Name!,
+                })
+        )
+
+        const allBuckets = await Promise.all(allBucketPromises)
+        const bucketsInRegion = _(allBuckets)
+            .reject(bucket => bucket.region !== this.regionCode)
+            .value()
+
+        return { buckets: bucketsInRegion }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public async listObjects(request: ListObjectsRequest): Promise<ListObjectsResponse> {
+        const s3 = await this.createS3()
+
+        let response: S3.ListObjectsV2Output
+        try {
+            response = await s3
+                .listObjectsV2({
+                    Bucket: request.bucketName,
+                    Delimiter: DEFAULT_DELIMITER,
+                    MaxKeys: request.maxResults ?? DEFAULT_MAX_KEYS,
+                    Prefix: request.folderPath,
+                    ContinuationToken: request.continuationToken,
+                })
+                .promise()
+        } catch (e) {
+            throw new S3Error(`Failed to list objects for bucket ${request.bucketName}: ${e}`)
+        }
+
+        const files: File[] = _(response.Contents)
+            .reject(file => file.Key === request.folderPath)
+            .map(
+                file =>
+                    new DefaultFile({
+                        key: file.Key!,
+                        partitionId: this.partitionId,
+                        bucketName: request.bucketName,
+                        lastModified: file.LastModified,
+                        sizeBytes: file.Size,
+                    })
+            )
+            .value()
+
+        const folders: Folder[] = _(response.CommonPrefixes)
+            .map(prefix => prefix.Prefix)
+            .compact()
+            .map(path => new DefaultFolder({ path, partitionId: this.partitionId, bucketName: request.bucketName }))
+            .value()
+
+        return {
+            files,
+            folders,
+            continuationToken: response.NextContinuationToken,
+        }
+    }
+
+    /**
+     * Looks up the region for the given bucket
+     *
+     * Note that although there is an S3#GetBucketLocation API,
+     * this is the suggested method of obtaining the region.
+     *
+     * @throws S3Error if there is an error calling S3.
+     */
+    private async lookupRegion(bucketName: string, s3: S3): Promise<string> {
+        try {
+            const response = await s3.headBucket({ Bucket: bucketName }).promise()
+            return response.$response.httpResponse.headers[BUCKET_REGION_HEADER]
+        } catch (e) {
+            throw new S3Error(`Failed to find region for bucket ${bucketName}: ${e}`)
+        }
+    }
+}
+
+export class DefaultBucket implements Bucket {
+    public readonly name: string
+    public readonly region: string
+    public readonly arn: string
+
+    public constructor({ partitionId, region, name }: { partitionId: string; region: string; name: string }) {
```

Comment:
really don't like that we're leaking `partition` all over the place - it should be conceptually a `Region` that has both a region ID (e.g. `us-west-2`) and a partition (e.g. `aws`)

Reply:
That's a fair point. The way it's modeled now reflects the `endpoints.json` file. The `partition` `has-many` `regions`. So the `regions` themselves have no "parent reference" back to their `partition`.

Do you think it's worth diverging from the `endpoints.json` format?